test/autolink_test.rb

```diff
@@ -279,23 +286,94 @@ def test_links_like_autolink_rails
     assert_linked '<a href="http://www.rubyonrails.com">Ruby On Rails</a>', '<a href="http://www.rubyonrails.com">Ruby On Rails</a>'
   end
 
-  if "".respond_to?(:force_encoding)
-    def test_copies_source_encoding
-      str = "http://www.bash.org"
+  def test_copies_source_encoding
+    str = "http://www.bash.org"
+
+    ret = Rinku.auto_link str
+    assert_equal str.encoding, ret.encoding
+
+    str.encode! 'binary'
+
+    ret = Rinku.auto_link str
+    assert_equal str.encoding, ret.encoding
+  end
+
+  def test_valid_encodings_are_generated
+    str = "&lt;a href=&#39;http://gi.co&#39;&gt;gi.co&lt;/a&gt;\xC2\xA0r"
+    assert_equal Encoding::UTF_8, str.encoding
+
+    res = Rinku.auto_link(str)
+    assert_equal Encoding::UTF_8, res.encoding
+    assert res.valid_encoding?
+  end
+
+  def test_polish_wikipedia_haha
+    url = "https://pl.wikipedia.org/wiki/Komisja_śledcza_do_zbadania_sprawy_zarzutu_nielegalnego_wywierania_wpływu_na_funkcjonariuszy_policji,_służb_specjalnych,_prokuratorów_i_osoby_pełniące_funkcje_w_organach_wymiaru_sprawiedliwości"
```

Comment:
I tried to pick up every substantial test case from the open issues. Thanks for reporting! 😄